File: js/src/containers/mp4.js
// Lazy ISO-BMFF atom walking. spec/10-containers.md §10.2
// Mirrors python/src/dashgps/containers/mp4.py

import { readSlab } from "../io.js";

const CONTAINERS = new Set(["moov", "trak", "mdia", "minf", "stbl", "udta", "moof", "traf", "mvex"]);
const FTYP_BRANDS = new Set(["ftyp", "styp", "moov", "free", "skip", "mdat", "wide", "pnot"]);

export class Atom {
  constructor(type, start, body, end, path) {
    this.type = type;
    this.start = start;
    this.body = body;
    this.end = end;
    this.path = path;
  }
  get bodySize() { return this.end - this.body; }
}

function typeStr(b, i) {
  return String.fromCharCode(b[i], b[i + 1], b[i + 2], b[i + 3]);
}

export async function looksLikeMp4(reader) {
  const n = reader.size();
  if (n < 16) return false;
  const head = await reader.readRange(0, 16);
  if (head.length < 16) return false;
  const size = ((head[0] << 24) >>> 0) + (head[1] << 16) + (head[2] << 8) + head[3];
  if (!FTYP_BRANDS.has(typeStr(head, 4))) return false;
  return size === 0 || size === 1 || (size >= 8 && size <= n);
}

export async function* iterAtoms(reader, start = 0, end, path = "", depth = 0,
                                 maxDepth = 6, maxAtoms = 4096) {
  const n = end === undefined ? reader.size() : end;
  let pos = start;
  let count = 0;
  while (pos + 8 <= n && count < maxAtoms) {
    const hdr = await readSlab(reader, pos, pos + 16);
    if (hdr.length < 8) return;
    let size = hdr.u32be(pos);
    const t = typeStr(hdr.data, 4);
    let body = pos + 8;
    if (size === 1) {
      if (hdr.length < 16) return;
      size = hdr.u64be(pos + 8);
      body = pos + 16;
    } else if (size === 0) {
      size = n - pos;
    }
    if (size < body - pos || pos + size > n) return;
    const aend = pos + size;
    const a = new Atom(t, pos, body, aend, path + "/" + t);
    yield a;
    count += 1;
    if (CONTAINERS.has(t) && depth < maxDepth) {
      yield* iterAtoms(reader, body, aend, a.path, depth + 1, maxDepth, maxAtoms);
    }
    pos = aend;
  }
}

File: python/src/dashgps/containers/mp4.py
"""Lazy ISO-BMFF (MP4/MOV) atom walking. spec/10-containers.md §10.2.

Never reads an atom's payload unless asked, so walking a 1 GB file costs a few dozen small reads.
"""

from ..io import read_slab

CONTAINERS = frozenset(
    (b"moov", b"trak", b"mdia", b"minf", b"stbl", b"udta", b"moof", b"traf", b"mvex")
)

FTYP_BRANDS = frozenset((b"ftyp", b"styp", b"moov", b"free", b"skip", b"mdat", b"wide", b"pnot"))


class Atom:
    __slots__ = ("type", "start", "body", "end", "path")

    def __init__(self, atype, start, body, end, path):
        self.type = atype
        self.start = start
        self.body = body
        self.end = end
        self.path = path

    @property
    def body_size(self):
        return self.end - self.body


def looks_like_mp4(reader):
    """Cheap structural check: a plausible top-level atom at offset 0."""
    n = reader.size()
    if n < 16:
        return False
    head = reader.read_range(0, 16)
    if len(head) < 16:
        return False
    size = (head[0] << 24) | (head[1] << 16) | (head[2] << 8) | head[3]
    atype = head[4:8]
    if atype not in FTYP_BRANDS:
        return False
    return size == 0 or size == 1 or (8 <= size <= n)


def iter_atoms(reader, start=0, end=None, path="", depth=0, max_depth=6, max_atoms=4096):
    """Yield Atom objects depth-first. Recurses into known container types."""
    n = reader.size() if end is None else end
    pos = start
    count = 0
    while pos + 8 <= n and count < max_atoms:
        hdr = read_slab(reader, pos, pos + 16)
        if len(hdr) < 8:
            return
        size = hdr.u32be(pos)
        atype = hdr.bytes(pos + 4, 4)
        body = pos + 8
        if size == 1:
            if len(hdr) < 16:
                return
            size = hdr.u64be(pos + 8)
            body = pos + 16
        elif size == 0:
            size = n - pos
        if size < (body - pos) or pos + size > n:
            return
        aend = pos + size
        a = Atom(atype, pos, body, aend, path + "/" + atype.decode("latin-1"))
        yield a
        count += 1
        if atype in CONTAINERS and depth < max_depth:
            for sub in iter_atoms(reader, body, aend, a.path, depth + 1, max_depth, max_atoms):
                yield sub
        pos = aend
